common/src/chain/config/mod.rs

```diff
@@ -44,6 +44,260 @@ pub enum ChainType {
     Signet,
 }
 
+impl ChainType {
+    const fn default_address_prefix(&self) -> &'static str {
+        match self {
+            ChainType::Mainnet => "mtc",
+            ChainType::Testnet => "tmt",
+            ChainType::Regtest => "rmt",
+            ChainType::Signet => "smt",
+        }
+    }
+
+    const fn default_magic_bytes(&self) -> [u8; 4] {
+        match self {
+            ChainType::Mainnet => [0x1a, 0x64, 0xe5, 0xf1],
+            ChainType::Testnet => [0x2b, 0x7e, 0x19, 0xf6],
+            ChainType::Regtest => [0xaa, 0xbb, 0xcc, 0xdd],
+            ChainType::Signet => [0xf3, 0xf7, 0x7b, 0x45],
+        }
+    }
+
+    fn default_genesis_init(&self) -> GenesisInit {
+        match self {
+            ChainType::Mainnet => GenesisInit::Mainnet,
+            ChainType::Testnet => todo!("Testnet genesis"),
+            ChainType::Regtest => GenesisInit::TEST,
+            ChainType::Signet => GenesisInit::TEST,
+        }
+    }
+
+    fn default_net_upgrades(&self) -> NetUpgrades<UpgradeVersion> {
+        match self {
+            ChainType::Mainnet => {
+                let pow_config = PoWChainConfig::new(*self);
+                let upgrades = vec![
+                    (
+                        BlockHeight::new(0),
+                        UpgradeVersion::ConsensusUpgrade(ConsensusUpgrade::IgnoreConsensus),
+                    ),
+                    (
+                        BlockHeight::new(1),
+                        UpgradeVersion::ConsensusUpgrade(ConsensusUpgrade::PoW {
+                            initial_difficulty: pow_config.limit().into(),
+                        }),
+                    ),
+                ];
+                NetUpgrades::initialize(upgrades).expect("net upgrades")
+            }
+            ChainType::Testnet => todo!("Testnet upgrades"),
+            ChainType::Regtest => NetUpgrades::unit_tests(),
+            ChainType::Signet => NetUpgrades::unit_tests(),
+        }
+    }
+}
+
+// Builder support types
+
+#[derive(Clone)]
+enum EmissionScheduleInit {
```

Comment:
All these are equivalent. I don't think it's a good idea to make them different arms of an enum that are processed differently, because this would increase the maintenance burden and we'll have different execution paths for different cases. Basically, Mainnet will have an execution path that will only be touched in mainnet... we're killing the benefit of functional tests (even though they're possibly built from the same tool, still). Whether you prefer Table or Fn, pick one, and convert everything to that in ChainConfig, and I honestly prefer Table because reading that as a variable in the debugger is possible.

Reply:
It all ends up as the `EmissionSchedule` type, which internally contains `Fn`. These are just different ways to initialise it with different tradeoffs that all get converted into it when you `.build()`. The real motivation here is `EmissionScheduleInit::Mainnet` which is a per-year emission schedule. In order to calculate the number of blocks per year, we need to know the final decision regarding the block interval and that is only available at the point `.build()` is called.

Reply:
I just realized this is going on only for the builder. I thought the enum is for the ChainConfig. My bad.

On another note, I'm not really happy about EmissionSchedule being an Fn... really... debugging that is going to be really nasty.